src/Arcus.WebApi.Security/Authentication/SharedAccessKeyAuthenticationAttribute.cs

```diff
@@ -0,0 +1,104 @@
+        /// </summary>
+        /// <param name="headerName">The name of the request header which value must match the stored secret.</param>
+        /// <param name="secretName">The name of the secret that's being retrieved using the <see cref="ISecretProvider.Get"/> call.</param>
+        public SharedAccessKeyAuthenticationAttribute(string headerName, string secretName)
+        {
+            if (String.IsNullOrWhiteSpace(headerName))
+            {
+                throw new ArgumentException("Header name cannot be blank", nameof(headerName));
+            }
+
+            if (String.IsNullOrWhiteSpace(secretName))
+            {
+                throw new ArgumentException("Secret name cannot be blank", nameof(secretName));
+            }
+
+            _headerName = headerName;
+            _secretName = secretName;
+        }
+        
+        /// <inheritdoc />
+        public override async Task OnActionExecutionAsync(ActionExecutingContext context, ActionExecutionDelegate next)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            if (context.HttpContext?.Request == null)
+            {
+                throw new ArgumentException(
+                    "Invalid action context given without any HTTP request", 
+                    nameof(context.HttpContext.Request));
+            }
+
+            if (context.HttpContext.Request.Headers == null)
+            {
+                throw new ArgumentException(
+                    "Invalid action context given without any HTTP request headers",
+                    nameof(context.HttpContext.Request.Headers));
+            }
+
+            if (context.HttpContext.RequestServices == null)
+            {
+                throw new ArgumentException(
+                    "Invalid action context given without any HTTP request services", 
+                    nameof(context.HttpContext.RequestServices));
+            }
+
+            var userDefinedSecretProvider = context.HttpContext.RequestServices.GetService<ISecretProvider>();
+            if (userDefinedSecretProvider == null)
+            {
+                throw new KeyNotFoundException(
+                    $"No configured {nameof(ISecretProvider)} implementation found in the request service container. "
+                    + "Please configure such an implementation (ex. in the Startup) of your application");
+            }
+
+            var cachedSecretProvider = new CachedSecretProvider(userDefinedSecretProvider);
+            string foundSecret = await cachedSecretProvider.Get(_secretName);
+            if (foundSecret == null)
+            {
+                throw new AuthenticationException(
+                    $"No secret found with name {_secretName} in {nameof(ISecretProvider)} implementation {userDefinedSecretProvider.GetType().Name}");
+            }
+
+            bool noRequestSecretHeaderFound = 
```

Comment:
Best to do this check before we get the secret, maybe we don't even need the secret as the header is not there?

Reply:
I think this is resolved, right @stijnmoreels ?

Reply:
Ah yes, is switched now to first look in the request headers.